[Pass Bug] 3D-rhs matmul carry (pl.create_tensor + if k0==0/else) emits 5D TLOAD that fails ISA layout assertion

### Pass Name

pto_backend AIC kernel emission (TLOAD layout selection for matmul rhs slice with leading-1 batch dim)

### Description

Following the pypto#1501 fix (which made the `pl.pipeline + pl.create_tensor + if kb_idx == 0: pl.matmul / else: pl.matmul_acc` carry pattern work for 2D matmul), the same fused-K-pipeline form on a **3D rhs** matmul still fails. The user has confirmed pypto#1467 (3D no-op batch-plane mat→mat tmov) is fixed locally; this is a distinct codegen issue exposed once that fix landed.

The rhs slice `w[g:g+1, n0:n0+N, k0:k0+K]` carries a leading-1 batch dimension. With the carry form, pypto lowers it into a `GlobalTensor` typed `Shape<1, 1, 1, N, K>` (three leading 1s — two extra padding axes beyond the source 3D shape) and emits a TLOAD into a 2D `Mat` tile. The pto-isa static-assert refuses that layout combination:

```
/data/czq/pto-isa/include/pto/npu/a2a3/TLoad.hpp:468:9: error: static assertion failed:
  Fix: TLOAD(MatTile, GlobalTensor) only support ND2ND/DN2DN/NZ2NZ/ND2NZ/DN2ZN!
```

(The full `pto::Tile<Mat, __bf16, 128, 128, ColMajor, ..., RowMajor, ...>` vs `GlobalTensor<__bf16, Shape<1, 1, 1, 128, 128>, Stride<...>, Layout::DN>` instantiation report is in the gist below.)

The peel-first-iter form (`pl.matmul` before the loop + `pl.matmul_acc` inside `pl.pipeline(1, ...)`) on the same 3D shape compiles fine — so the breakage is specific to the carry pattern's lowering of the rhs slice.

A second, related failure mode appears when the `pl.create_tensor` for the acc is **hoisted out of `pl.at`**: ptoas reports

```
'pto.tload' op expects A2/A3 tload dst to use loc=vec or loc=mat
```

pypto emits a TLOAD of the GM acc tensor's preload into the cube accumulator (`loc=acc`), which the ISA does not support. Same root cause family, different surface.

Real-world impact: `models/deepseek/v4/decode_sparse_attn.py` proj_a (`wo_a` is 3D `[O_GROUPS, O_LORA, O_GROUP_IN]`) must stay in peel-first-iter form; `qr_proj_matmul`-style carry rewrite was attempted twice and hits this.

### Git Commit ID

9ff54ef2366ab85ebaa485fdc7e808732cdbddf1 (pypto, with user's local pypto#1467 fix applied)

### Before IR (Input)

```python
import pypto.language as pl

T = 128
N = 128
K = 128
G = 8
NB = 4
TOTAL_K = 2048


@pl.jit
def matmul_3d_carry(
    x: pl.Tensor[[T, TOTAL_K], pl.BF16],
    w: pl.Tensor[[G, N * NB, TOTAL_K], pl.BF16],
    out: pl.Out[pl.Tensor[[G, T, N * NB], pl.FP32]],
):
    for g in pl.parallel(0, G, 1):
        for nb in pl.parallel(0, NB, 1):
            n0 = nb * N
            with pl.at(level=pl.Level.CORE_GROUP, name_hint="matmul_3d"):
                acc = pl.create_tensor([1, T, N], dtype=pl.FP32)
                for kb in pl.pipeline(0, TOTAL_K // K, stage=2):
                    k0 = kb * K
                    x_chunk = x[:, k0:k0 + K]
                    w_chunk = w[g:g + 1, n0:n0 + N, k0:k0 + K]   # 3D rhs with leading-1
                    if k0 == 0:
                        acc = pl.matmul(x_chunk, w_chunk, b_trans=True, out_dtype=pl.FP32)
                    else:
                        acc = pl.matmul_acc(acc, x_chunk, w_chunk, b_trans=True)
                out[g:g + 1, :, n0:n0 + N] = acc
    return out
```

### Expected IR (After Transformation)

The carry form should compile and run equivalently to the peel-first-iter form below (which compiles fine on the same 3D shape, demonstrating the lowering path is otherwise viable). The rhs `GlobalTensor` should keep its source 3D shape — no extra leading-1 padding to 5D — and the cube-side TLOAD should match an ISA-supported layout pair.

```python
@pl.jit
def matmul_3d_peel(
    x: pl.Tensor[[T, TOTAL_K], pl.BF16],
    w: pl.Tensor[[G, N * NB, TOTAL_K], pl.BF16],
    out: pl.Out[pl.Tensor[[G, T, N * NB], pl.FP32]],
):
    for g in pl.parallel(0, G, 1):
        for nb in pl.parallel(0, NB, 1):
            n0 = nb * N
            with pl.at(level=pl.Level.CORE_GROUP, name_hint="matmul_3d_peel"):
                x0 = x[:, 0:K]
                w0 = w[g:g + 1, n0:n0 + N, 0:K]
                acc = pl.matmul(x0, w0, b_trans=True, out_dtype=pl.FP32)
                for kb in pl.pipeline(1, TOTAL_K // K, stage=2):
                    k0 = kb * K
                    x_chunk = x[:, k0:k0 + K]
                    w_chunk = w[g:g + 1, n0:n0 + N, k0:k0 + K]
                    acc = pl.matmul_acc(acc, x_chunk, w_chunk, b_trans=True)
                out[g:g + 1, :, n0:n0 + N] = acc
    return out
```

### Actual IR or Error

Failure mode (a) — `pl.create_tensor` inside `pl.at` (Python AIC kernel cpp instantiation):

```text
/data/czq/pto-isa/include/pto/npu/a2a3/TLoad.hpp:468:9: error: static assertion failed
  due to requirement 'isSameLayout || ...':
  Fix: TLOAD(MatTile, GlobalTensor) only support ND2ND/DN2DN/NZ2NZ/ND2NZ/DN2ZN!

note: in instantiation of function template specialization
  'pto::TLOAD<pto::Tile<pto::TileType::Mat, __bf16, 128, 128, pto::BLayout::ColMajor, -1, -1,
              pto::SLayout::RowMajor, 512, pto::PadValue::Null, pto::CompactMode::Null>,
              pto::GlobalTensor<__bf16, pto::Shape<1, 1, 1, 128, 128>,
                                pto::Stride<1048576, 1048576, 1048576, 1, 2048>,
                                pto::Layout::DN>>'
  requested here
    TLOAD(v30, v34);
```

The `GlobalTensor` is `Shape<1, 1, 1, 128, 128>` (5D) for a source slice that is 3D `[1, 128, 128]` — two extra leading 1s.

Failure mode (b) — `pl.create_tensor` hoisted just before `pl.at` (ptoas):

```text
ptoas compilation failed:
  error: 'pto.tload' op expects A2/A3 tload dst to use loc=vec or loc=mat
```

The corresponding `.pto` excerpt shows `pto.tload` with `outs(... loc=acc, dtype=f32, rows=128, cols=128 ...)` — preloading the create_tensor's GM acc directly into the cube accumulator.

### NPU Kind (if relevant)

Ascend 910B (Atlas A2/A3)

### Host Platform (if relevant)

Linux (aarch64)

### Additional Context

- Standalone repro at `models/deepseek/v4/repro_3d_matmul_carry_draft.py` in `pypto-lib`. `python models/deepseek/v4/repro_3d_matmul_carry_draft.py -p a2a3` reproduces failure mode (a). Flipping `ACC_INSIDE_PL_AT = False` reproduces failure mode (b).
- The `matmul_3d_peel` function in the same file is the working peel-first-iter form, used as the Expected baseline.
- Related: pypto#1501 (closed, 2D carry), pypto#1467 (3D mat→mat tmov, user-side fixed).
- Affected production kernel: `models/deepseek/v4/decode_sparse_attn.py` proj_a scope (commit `2f5f10e8` in pypto-lib). The `qr_proj_matmul` style refactor was attempted but reverted to peel-first-iter form pending this fix.

--- a/src/ir/transforms/flatten_tile_nd_to_2d_pass.cpp
+++ b/src/ir/transforms/flatten_tile_nd_to_2d_pass.cpp
@@ -1622,9 +1622,21 @@ std::vector<StmtPtr> TransformBody(const std::vector<StmtPtr>& stmts, FlattenCon
         auto [merged, last] = ComputeMergedShape(result_tile->shape_, "tile.load result");
 
         auto flat_shape_exprs = Make2DShapeExprs(merged, last, span);
-        // Assign the implicit TileView for the flattened 2D shape+memory_space.
-        auto flat_tile_view = std::make_optional(
-            tile_view_semantics::GetImplicitTileView(flat_shape_exprs, result_tile->memory_space_));
+        // Preserve any TileView (blayout/slayout/fractal/pad) the source tile
+        // already carried — e.g. LowerCompositeOps tags a transposed-load Mat
+        // rhs with TileView(blayout=row_major, slayout=col_major) so the
+        // downstream TLOAD matches the DN2ZN pattern. The implicit Mat default
+        // (col/row = ND) would otherwise emit an unsupported DN2ND TLOAD
+        // (#1540). Mirrors ExtractBatchPage Strategy 1 above.
+        std::optional<TileView> flat_tile_view;
+        if (result_tile->tile_view_.has_value()) {
+          const auto& orig_tv = *result_tile->tile_view_;
+          flat_tile_view = TileView(flat_shape_exprs, /*stride=*/{}, /*start_offset=*/nullptr,
+                                    orig_tv.blayout, orig_tv.slayout, orig_tv.fractal, orig_tv.pad);
+        } else {
+          flat_tile_view =
+              tile_view_semantics::GetImplicitTileView(flat_shape_exprs, result_tile->memory_space_);
+        }
         auto flat_tile_type = std::make_shared<TileType>(flat_shape_exprs, result_tile->dtype_, std::nullopt,
                                                          flat_tile_view, result_tile->memory_space_);
         auto flat_call =

--- a/tests/ut/ir/transforms/test_flatten_tile_nd_to_2d.py
+++ b/tests/ut/ir/transforms/test_flatten_tile_nd_to_2d.py
@@ -1627,6 +1627,103 @@ class TestFlattenTileNdTo2DBatchMatmul:
         # that Strategy 1 produces.
         assert op_names == ["tile.load", "tile.load", "tile.matmul", "tile.store"]
 
+    def test_rank3_mat_load_under_if_preserves_explicit_tile_view(self):
+        """Regression for #1540: a rank>2 ``tile.load`` whose downstream
+        ``tile.batch_matmul`` use is hidden inside an ``if/else`` block must
+        still carry its explicit ``TileView`` (blayout=row_major,
+        slayout=col_major) onto the flattened 2D load.
+
+        The pre-scan at the top of ``TransformBody`` walks only top-level
+        statements, so when the matmul lives inside an ``IfStmt`` body the
+        load is not added to ``batch_matmul_only_vars`` and Strategy 1 cannot
+        re-emit per-batch loads. The load instead takes the fallback rewrite
+        path. Before #1540 that path computed a fresh implicit ``TileView``
+        from (shape, memory_space), clobbering the upstream NZ-layout
+        annotation that ``LowerCompositeOps`` set on transposed-load Mat rhs
+        operands. Downstream codegen then emitted ``pto.tload DN→ND``, which
+        ``pto-isa`` rejects.
+        """
+        T, K, N = 16, 128, 64
+
+        @pl.program
+        class Before:
+            @pl.function(type=pl.FunctionType.InCore)
+            def main_incore_0(
+                self,
+                h: pl.Tensor[[T, K], pl.BF16],
+                w: pl.Tensor[[1, N, K], pl.BF16],
+                cond: pl.Scalar[pl.INDEX],
+                out_0: pl.Out[pl.Tensor[[1, T, N], pl.FP32]],
+            ) -> pl.Tensor[[1, T, N], pl.FP32]:
+                lhs: pl.Tile[[T, K], pl.BF16, pl.Mem.Mat] = pl.tile.load(
+                    h, [0, 0], [T, K], target_memory=pl.Mem.Mat
+                )
+                # Explicit NZ-layout annotation — matches what LowerCompositeOps
+                # emits for transposed-load Mat rhs operands of pl.matmul.
+                # transpose=True swaps the last two dims: source slice [1, N, K]
+                # becomes tile [1, K, N].
+                rhs: pl.Tile[
+                    [1, K, N],
+                    pl.BF16,
+                    pl.Mem.Mat,
+                    pl.TileView(
+                        blayout=pl.TileLayout.row_major,
+                        slayout=pl.TileLayout.col_major,
+                    ),
+                ] = pl.tile.load(w, [0, 0, 0], [1, N, K], target_memory=pl.Mem.Mat, transpose=True)
+                # The use lives inside an if/else; the pre-scan does not see it,
+                # so the fallback rewrite path runs on ``rhs``.
+                if cond == 0:
+                    mm0: pl.Tile[[1, T, N], pl.FP32] = pl.tile.batch_matmul(lhs, rhs)
+                    out_tile = pl.yield_(mm0)
+                else:
+                    mm1: pl.Tile[[1, T, N], pl.FP32] = pl.tile.batch_matmul(lhs, rhs)
+                    out_tile = pl.yield_(mm1)
+                out_0 = pl.tile.store(out_tile, [0, 0, 0], out_0)
+                return out_0
+
+            @pl.function
+            def main(
+                self,
+                h: pl.Tensor[[T, K], pl.BF16],
+                w: pl.Tensor[[1, N, K], pl.BF16],
+                cond: pl.Scalar[pl.INDEX],
+            ) -> pl.Tensor[[1, T, N], pl.FP32]:
+                out_0 = pl.create_tensor([1, T, N], dtype=pl.FP32)
+                return self.main_incore_0(h, w, cond, out_0)
+
+        After = passes.flatten_tile_nd_to_2d()(Before)
+        after_func = After.get_function("main_incore_0")
+        assert after_func is not None
+
+        # Locate the rhs load — the only ``tile.load`` with ``transpose=True``.
+        rhs_loads = [
+            stmt
+            for stmt in cast(ir.SeqStmts, after_func.body).stmts
+            if isinstance(stmt, ir.AssignStmt)
+            and isinstance(stmt.value, ir.Call)
+            and stmt.value.op.name == "tile.load"
+            and stmt.value.kwargs.get("transpose") is True
+        ]
+        assert len(rhs_loads) == 1, (
+            f"expected exactly one transposed rhs tile.load after flatten, got {len(rhs_loads)}"
+        )
+        rhs_load = rhs_loads[0]
+        result_type = cast(ir.TileType, rhs_load.value.type)
+
+        # Result must be 2D (rank>2 was the input).
+        assert len(result_type.shape) == 2
+
+        # Use the effective view to compare layouts robustly against
+        # canonicalization (implicit views collapse to ``tile_view is None``).
+        eff = result_type.get_effective_tile_view()
+        assert eff.blayout == ir.TileLayout.row_major, (
+            f"flattened rhs Mat tile lost NZ blayout (#1540): blayout={eff.blayout}, slayout={eff.slayout}"
+        )
+        assert eff.slayout == ir.TileLayout.col_major, (
+            f"flattened rhs Mat tile lost NZ slayout (#1540): blayout={eff.blayout}, slayout={eff.slayout}"
+        )
+
 
 # ----------------------------------------------------------------------------
 # tile.batch_matmul_acc lowering

